File: 2560-house-robber-iv/2560-house-robber-iv.java
class Solution {
    public int minCapability(int[] nums, int k) {
        int minCap = 1;
        int maxCap=nums[0];
        for(int i=0;i<nums.length;i++){
            maxCap = Math.max(maxCap,nums[i]);
        }
        int len = nums.length;
        
        
        while(minCap < maxCap ){
            int midCap = ( maxCap + minCap )/2 ;
            int total_possible = 0;

            for(int i=0;i<len;++i){
                if(nums[i]<=midCap){
                    total_possible++;
                    i++;
                }       
            }

            if(total_possible>=k)
                maxCap =midCap;
            else 
                minCap =midCap+1;
        }
        
        return minCap;
    }
}

File: 2560-house-robber-iv/2560-house-robber-iv.py
class Solution:
    def minCapability(self, nums, k):
        def canRob(cap):
            count, i, n = 0, 0, len(nums)

            while i < n:
                if nums[i] <= cap:
                    count += 1
                    i += 2
                else:
                    i += 1
            return count >= k

        lo, hi = min(nums), max(nums)

        while lo < hi:
            mid = (lo + hi) // 2
            if canRob(mid):
                hi = mid
            else:
                lo = mid + 1
        return lo
